compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/DerivedConfigUtils.cpp

```diff
@@ -54,16 +54,30 @@ getThreadTileSizesFromLoopRanges(SmallVector<int64_t> loopRanges,
   }
 
   SmallVector<int64_t> tileSizes(loopRanges.size(), 0);
-  tileSizes.back() = vectorSize;
-  int64_t residualNumThreads = numThreads / (loopRanges.back() / vectorSize);
-  for (int i = tileSizes.size() - 2, e = 0; i >= e; --i) {
+  int64_t residualNumThreads = numThreads;
+  int i = tileSizes.size() - 1;
+  // Set as many inner tile sizes to 0 as possible to use the full vectorSize.
+  while (i >= 0) {
+    if (loopRanges[i] > vectorSize) {
+      tileSizes[i] = vectorSize;
+      residualNumThreads = numThreads / (loopRanges.back() / vectorSize);
+      i--;
+      break;
+    }
+    tileSizes[i] = 0;
+    vectorSize /= loopRanges[i];
+    i--;
+  }
+  // Set as many remaining tile sizes to 1 as possible to use all threads.
+  while (i >= 0) {
     if (loopRanges[i] >= residualNumThreads) {
       tileSizes[i] = loopRanges[i] / residualNumThreads;
       residualNumThreads = 1;
       break;
     }
     tileSizes[i] = 1;
     residualNumThreads /= loopRanges[i];
+    i--;
```

Comment:
What is this change for? I don't see any tests for it and it's not clear to me why inner tile sizes would be 0

Reply:
+1

Reply:
I added a test for it now. The case is something like `linalg.copy ... -> tensor<4x16x8x4x16x2x4xf16>`. The preferred load width is `8xf16`, but the inner size is only `4xf16`. This will make the tile sizes become 0 for the inner 2 dimensions, which makes the inner load shape fill the full vector size.

This is somewhat hopeful that the inner dimensions will be contiguous, which is not always the case. My understanding , though, was that loads will not be coalesced unless the preferred load bitwidth is used anyway, so this should be worth doing.

Reply:
no, coalescing can happen for any dword * pow2 load width I believe. But if the inner most dim is non-contiguous we can't coalesce anyway so I suppose this is fine.